### PR TITLE
Debian package

### PR DESCRIPTION
**Pull request by DmitryZabolotsky**
Includes user and directories creation for Consensus node

Signed-off-by: Dmitriy Zabolotskiy <d.zabolotskiy@yadro.com>

### CODE REVIEW

File: debian/control [hunk 1]
```diff
@@ -0,0 +1,25 @@
+Source: neo-go
+Section: misc
+Priority: optional
+Maintainer: NeoSPCC <tech@nspcc.ru>
+Build-Depends: debhelper-compat (= 13), git, devscripts
+Standards-Version: 4.5.1
+Homepage: https://fs.neo.org/
```
Comment (AnnaShaleva) [inline debian/control:7]: Is it our (neo-go) homepage?

File: debian/control [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+Vcs-Git: https://github.com/nspcc-dev/neo-go.git
+Vcs-Browser: https://github.com/nspcc-dev/neo-go
+
+Package: neo-go
+Architecture: any
+Depends: ${misc:Depends}
+Description: This project aims to be a full port of the original C# Neo project
```
Comment (AnnaShaleva) [inline debian/control:14]: ... full port including some extensions?

File: debian/control [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ A complete toolkit for the NEO blockchain, including
+ .
+  - Consensus node
+  - RPC node & client
+  - CLI tool
+  - Smart contract compiler
+  - NEO virtual machine
+  - Smart contract examples
+  - Oracle service
+  - State validation service
```
Comment (AnnaShaleva) [inline debian/control:24]: Either these services should be excluded from the list (they are a part of the node) or Notary service should be added.

File: debian/neo-go.docs [hunk 4]
```diff
@@ -0,0 +1,4 @@
+docs/*
+CONTRIBUTING.md
+ROADMAP.md
+README.md
```
Comment (AnnaShaleva) [inline debian/neo-go.docs:4]: Add `pkg/rpcclient/doc.go` (and `pkg/rpcclient/doc_test` to examples section)? An important one.

File: debian/neo-go.postinst [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+#          <failed-install-package> <version> `removing'
+#          <conflicting-package> <version>
+# for details, see https://www.debian.org/doc/debian-policy/ or
+# the debian-policy package
+
+
+case "$1" in
+    configure)
+        USERNAME=neo-go
+        id -u $USERNAME >/dev/null 2>&1 || useradd -s /usr/sbin/nologin -d /var/lib/$USERNAME --system -M -U -c "Neo-go consensus node" neo-go
```
Comment (AnnaShaleva) [inline debian/neo-go.postinst:24]: Nitpicking, but it's not always the consensus node. E.g. it can be just an RPC node.

File: debian/neo-go.install [hunk 6]
```diff
@@ -0,0 +1,4 @@
+config/protocol.mainnet.yml etc/neo-go
+config/protocol.privnet.yml etc/neo-go
+config/protocol.testnet.yml etc/neo-go
```
Comment (AnnaShaleva) [inline debian/neo-go.install:3]: We also have a configuration for neofs storage node (mainnet/testnet). Although we don't have a separate CLI command to start the node with this config, it can still be useful for someone.

**Review by AnnaShaleva — COMMENTED** [review]
I've reviewed the debpackage part itself and it seems to be OK for me, but I need someone more qualified to check it one more time.

File: debian/control (hunk 2 above)
Comment (roman-khimov) [reply debian/control:14]: It's copy-pasted from our README, so it's OK for now, but maybe we want to review/change that.

File: debian/neo-go.docs (hunk 4 above)
Comment (roman-khimov) [reply debian/neo-go.docs:4]: What's packaged here is the neo-go binary that implements a node, a compiler and all of these things. `rpcclient` is a programmer-facing API, it's somewhat different and should not be packaged as a document (it's Go code after all).

File: .gitignore [hunk 7]
```diff
@@ -28,7 +28,7 @@ bin/
 *~
 TAGS
 
-# storage 
+# storage
```
Comment (roman-khimov) [inline .gitignore:31]: Not a useful change, can be omitted or moved into a separate commit.

File: Makefile [hunk 8]
```diff
@@ -22,6 +22,13 @@ BUILD_FLAGS = "-X '$(REPO)/pkg/config.Version=$(VERSION)' -X '$(REPO)/cli/smartc
 
 IMAGE_REPO=nspccdev/neo-go
 
+# .deb package versioning
+OS_RELEASE = $(shell lsb_release -cs)
+PKG_VERSION ?= $(shell echo $(VERSION) | sed "s/^v//" | \
+			sed -E "s/(.*)-(g[a-fA-F0-9]{6,8})(.*)/\1\3~\2/" | \
```
Comment (roman-khimov) [inline Makefile:28]: Why this one?

File: debian/neo-go.dirs [hunk 9]
```diff
@@ -0,0 +1,2 @@
+var/lib/neo-go
+etc/neo-go
```
Comment (roman-khimov) [inline debian/neo-go.dirs:2]: We need to fix `cli/options/options.go`, move `configPath` into `pkg/config` maybe and set it to `/etc/neo-go` for packaged neo-go.

File: debian/neo-go.postrm [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+#        * <new-postrm> `abort-upgrade' <old-version>
+#        * <disappearer's-postrm> `disappear' <overwriter>
+#          <overwriter-version>
+# for details, see https://www.debian.org/doc/debian-policy/ or
+# the debian-policy package
+
+
+case "$1" in
+    purge
+        rm -rf /var/lib/neofs/neo-go
```
Comment (roman-khimov) [inline debian/neo-go.postrm:24]: `var/lib/neo-go`?

File: debian/neo-go.prerm [hunk 11]
```diff
@@ -0,0 +1,38 @@
+#!/bin/sh
+# prerm script for neo-go
```
Comment (roman-khimov) [inline debian/neo-go.prerm:2]: And this one.

File: debian/neo-go.service [hunk 12]
```diff
@@ -0,0 +1,17 @@
+[Unit]
+Description=Neo Go node
```
Comment (roman-khimov) [inline debian/neo-go.service:2]: It's either `NeoGo` or `neo-go`.

File: docs/building-deb-package.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+* `make all`
+
+## Installing packaging dependencies
+
+```shell
+$ sudo apt install debhelper-compat dh-sequence-bash-completion devscripts
+```
+
+Warining: number of package installed is pretty large considering dependecies.
```
Comment (roman-khimov) [inline docs/building-deb-package.md:18]: dependencies

File: docs/building-deb-package.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+# Package versioning
+
+By default, package version is based on product version and may also contain git
+tags and hashes.
+
+Package version could be overwritten by setting `PKG_VERSION` variable before
+build, Debian package versioning rules should be respected.
+
+```shell
+$ PKG_VERSION=0.32.0 make debpackge
```
Comment (roman-khimov) [inline docs/building-deb-package.md:45]: debpackage

File: debian/control (hunk 2 above)
Comment (AnnaShaleva) [reply debian/control:14]: See https://github.com/nspcc-dev/neo-go/issues/2779.

File: debian/neo-go.dirs (hunk 9 above)
Comment (DmitryZabolotsky) [reply debian/neo-go.dirs:2]: Will it be more suitable if moved to /etc/neofs/neo-go and /var/lib/neofs/neo-go accordingly?
both directories are defined in Unit file anyways, like:
...
ExecStart=/usr/bin/neo-go node --mainnet --config-path=/etc/neofs/neo-go/
...
WorkingDirectory=/var/lib/neofs/neo-go

File: debian/control (hunk 3 above)
Comment (DmitryZabolotsky) [reply debian/control:24]: Section from README file was copied here from https://github.com/nspcc-dev/neo-go
I'm trying not to make any changes to information present

File: debian/neo-go.install (hunk 6 above)
Comment (DmitryZabolotsky) [reply debian/neo-go.install:3]: All configs present in source code are copied to examples directory for reference.
These 3 files were installed in original Makefile so again here I'm trying to introduce as little changes as possible.

File: debian/neo-go.postinst (hunk 5 above)
Comment (DmitryZabolotsky) [reply debian/neo-go.postinst:24]: Service file is installed in inactive and stopped state, so it should not break anything.
Probably, having different unit files for different operation modes could be usefull?
If so I'll add them.

File: .gitignore (hunk 7 above)
Comment (DmitryZabolotsky) [reply .gitignore:31]: Sorry it's auto whitespace trimming function :-( I'll take care of it

File: debian/neo-go.dirs (hunk 9 above)
Comment (roman-khimov) [reply debian/neo-go.dirs:2]: I don't think so, NeoGo is frequently used as a standalone package without NeoFS, so `/etc/neo-go` and `/var/lib/neo-go` are correct directories for it.

File: debian/neo-go.postinst (hunk 5 above)
Comment (roman-khimov) [reply debian/neo-go.postinst:24]: It doesn't need additional unit files, consensus/RPC/any other service operation only depends on configuration. But the comment for the user is incorrect, is should be something like "NeoGo node" and that's it.

File: Makefile (hunk 8 above)
Comment (DmitryZabolotsky) [reply Makefile:28]: It allows to include git hash aside from tag to allow developers builds be easily differentiated and sorted via apt.
Alas, Debian rules for package versioning are pretty restricted so various "hacks" are used here.

File: debian/neo-go.prerm (hunk 11 above)
Comment (DmitryZabolotsky) [reply debian/neo-go.prerm:2]: 4 files make full templates set for service lifecycle.
Never considered them useless but I'll gladly remove these 2.

**Comment by roman-khimov** [comment]
Looks like this attempt has failed, unfortunately.
